File: python/MedianofTwoSortedArrays.py
class Solution:
    # @return a float
    def findAK(self, A, B, k):
        left, right = -1, len(A)-1
        while left < right -1 :
            mid = left + (right - left) / 2
            p1 = k - mid - 1
            if p1 < 0 :
                right = mid
            elif p1 > len(B) :
                left = mid
            else :
                v = A[mid] if p1 == 0 else B[p1-1]
                if A[mid] >= v :
                    right = mid
                else :
                    left = mid
        p1 = k - right - 1
        if p1 < 0 or p1 > len(B) or right < 0 : return -1
        a1 = A[right] if p1 == 0 else B[p1-1]
        a2 = A[right] if p1 == len(B) else B[p1]
        return right if a1 <= A[right] <= a2 else -1
    def findK(self, A, B, k) :
        v = self.findAK(A, B, k)
        if v != -1 :
            return A[v]
        else :
            return B[self.findAK(B, A, k)]
    def findMedianSortedArrays(self, A, B):
        LEN = len(A) + len(B)
        if (LEN%2) :
            return self.findK(A, B, LEN / 2+1) * 1.0
        else :
            v = self.findK(A, B, LEN/2) * 1.0
            v2 = self.findK(A, B, LEN/2+1)
            return (v+v2)/2




sol = Solution()
a1 = [1,3,5,7,9]
a2 = [2,4,6,8,10]
print sol.findAK(a1, a2, 5)
print sol.findAK(a1, a2, 6)
print sol.findMedianSortedArrays([], [1])
print sol.findMedianSortedArrays([2], [])

print sol.findMedianSortedArrays([], [2,3])

File: cpluplus/Median_of_Two_Sorted_Arrays.cc
/*
 * author: yuandx
 * create: 2015-09-30
 * email: yuandx@mvad.com
 */

#include <iostream>
#include <vector>
#include <map>
#include <set>
#include <string>
#include <utility>
#include <functional>
#include <algorithm>
#include <stdio.h>
#include <climits>
using std::cout;
using std::cin;
using std::endl;
using std::vector;
using std::map;
using std::set;
using std::pair;
using std::sort;
using std::string;
class Solution
{
public:
  /*
   * mid, k-1-mid, k-2-mid
   */
  bool FindKInV1(vector<int>& v1, vector<int>& v2, int k, int &v)
  {
    int size1 = static_cast<int>(v1.size()), size2 = static_cast<int>(v2.size());
    int left = 0, right = size1 - 1;
    while (left <= right)
    {
      int mid = (left + right) / 2;
      cout << mid << endl;
      int v_length = k - mid - 1; //v1 is length
      if (v_length < 0)
      {
        right = mid - 1;
      }
      else if (v_length > size2)
      {
        left = mid + 1;
      }
      else
      {
        int left_v = v_length == 0 ? INT_MIN : v2[v_length - 1];
        int right_v = v_length == size2 ? INT_MAX : v2[v_length];
        if (v1[mid] < left_v)
        {
          left = mid + 1;
        }
        else if (v1[mid] > right_v)
        {
          right = mid - 1;
        }
        else
        {
          v = v1[mid];
          return true;
        }
      }
    }
    return false;
  }
  double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2)
  {
    int size = static_cast<int>(nums1.size() + nums2.size());
    if ((size & 1) == 1)
    {
      int v, K = size / 2 + 1;
      FindKInV1(nums1, nums2, K, v) || FindKInV1(nums2, nums1, K, v);
      return v;
    }
    else
    {
      int v1 = 0, K1 = size / 2;
      int v2 = 0, K2 = size / 2 + 1;
      FindKInV1(nums1, nums2, K1, v1) || FindKInV1(nums2, nums1, K1, v1);
      FindKInV1(nums1, nums2, K2, v2) || FindKInV1(nums2, nums1, K2, v2);
      cout << v1 << "\t" << v2 << endl;
      return (v1 + v2) / 2.0;
    }
  }
};

int main()
{
  Solution sol;
  vector<int> v1 = {1, 2}, v2 = v1;
  cout << sol.findMedianSortedArrays(v1, v2) << endl;
  return 0;
}
